### PR TITLE
Updated table component

### LINKED ISSUE #15018: Build out custom Table components

### PR DESCRIPTION
**Pull request by bbovenzi**
Custom table component to use across the new UI when a table is needed. It uses [react-table](https://react-table.tanstack.com/) for handling table logic (pagination, sorting, rendering) and chakra components for styling.

Closes: #15018

To use the table component you need to specify `data` and `columns`. Both need to be memo-ized for performance.

Data can be raw numbers, strings, boolean values or can accept a react component.
Options for columns can be found [here](https://react-table.tanstack.com/docs/api/useTable#column-options)

Client-side sorting and pagination work without any config, but server-side will require extra props to be sent to the `<Table />` component

Also included is an improved loading state with placeholder rows instead of simple "Loading..." text:

Screen Shot 2021-05-12 at 2 46 01 PM

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: airflow/ui/src/interfaces/react-table-config.d.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+  UseSortByColumnOptions,
+  UseSortByColumnProps,
+  UseSortByHooks,
+  UseSortByInstanceProps,
+  UseSortByOptions,
+  UseSortByState,
+} from 'react-table';
+
+declare module 'react-table' {
+  // take this file as-is, or comment out the sections that don't apply to your plugin configuration
```
Comment (ashb) [inline airflow/ui/src/interfaces/react-table-config.d.ts]: Left over scaffolding comment?
```suggestion
```

File: airflow/ui/src/components/Table.tsx [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+                  >
+                    {cell.render('Cell')}
+                  </Td>
+                ))}
+              </Tr>
+            );
+          })}
+        </Tbody>
+      </ChakraTable>
+      <Flex alignItems="center" justifyContent="flex-end">
```
Comment (ryanahamilton) [inline airflow/ui/src/components/Table.tsx]: The pagination gets kind of lost aligned to the right. I'd suggest aligning to the left and adding some margin.
```suggestion
      <Flex alignItems="center" justifyContent="flex-start" my={4}>
```

File: airflow/ui/src/components/Table.tsx [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        <IconButton variant="ghost" onClick={handlePrevious} disabled={!canPreviousPage} aria-label="Previous Page">
+          <MdKeyboardArrowLeft />
+        </IconButton>
+        <IconButton variant="ghost" onClick={handleNext} disabled={!canNextPage} aria-label="Next Page">
+          <MdKeyboardArrowRight />
+        </IconButton>
+        <Text>
+          {pageIndex + 1}
+          {' of '}
+          {pageCount}
```
Comment (ryanahamilton) [inline airflow/ui/src/components/Table.tsx]: I think the pagination label could be improved if we gave more context. e.g. change "1 of 2" to "1-24 of 36 DAGs"

File: airflow/ui/src/components/Table.tsx [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    <>
+      <ChakraTable {...getTableProps()}>
+        <Thead>
+          <Tr>
+            {allColumns.map((column) => (
+              <Th
+                {...column.getHeaderProps(column.getSortByToggleProps())}
+              >
+                {column.render('Header')}
+                <chakra.span pl="2">
```
Comment (ryanahamilton) [inline airflow/ui/src/components/Table.tsx]: Can probably drop this span/padding—it looks a bit too distant from the heading text with it.

**Review by ryanahamilton — COMMENTED** [review]
Overall this is really great and well organized. A couple general comments/questions:

- Should the column sort arrow indicator in the `th` be shown by default? Seems like it might be useful to indicate the current state and also improve discoverability of the feature. That would probably also alleviate the visual jumping occurring when the sort icon appears/disappears.
- Have you thought about representation of sort state in the URL params at all? I don't think it needs to be included in this PR, but will certainly be a feature we'd want (for all table implementations). I'm curious how much additional complexity it would add?
- Very cool to see the skeleton rows working. We'll need to work on refining a bit (later). The table column jumping that occurs when switching to the real content row kind of negates the initial goal of the skeletons.

**Comment by bbovenzi** [comment]
- Added an icon to show which columns are sortable:
Screen Shot 2021-05-26 at 1 08 41 PM
- Updated pagination as suggested
- I haven't been able to get the loader to render the same way but for sure something we should improve eventually
- For client-side automatic sorting and filtering we can have the table component manage the params, but I think anything server-side will have to be managed by it's parent component. Can certainly be the next PR.

**Review by ryanahamilton — APPROVED** [review]
Nice work!

**Pull request merged**